### PR TITLE
Add support for worker persistence with KEDA v2.0.0 in helm chart

### PR DESCRIPTION
**Pull request by dstandish**
Here we upgrade keda to v2.0.0 in helm chart update the docs since things have changed a bit.

Notably, since 2.0.0 adds support for statefulsets, the ScaledObject definition is updated to add support for the case where `workers.persistence.enabled=True`.

### CODE REVIEW

**Comment by dstandish** [comment]
@mik-laj following your comment here https://github.com/apache/airflow/pull/13183#issuecomment-748549890 i added a mechanism for validating CRD objects

i enabled schema validation for the keda scaledojbect, and also removed the option to disable schema validation since it's no longer necessary

File: chart/README.md
Comment (ianstanton) [inline chart/README.md]:
```suggestion
Enable for the airflow instance by setting `workers.keda.enabled=true` in your helm command or in the `values.yaml`.
```

**Review by ianstanton — CHANGES_REQUESTED** [review]
@dstandish Minor fix requested in README

**Comment by mik-laj** [comment]
Can you do a rebase?

**Review by mik-laj — APPROVED** [review]
LGTM

**Comment by dstandish** [comment]
Great -- will do

File: docs/helm-chart/keda.rst
Comment (dstandish) [inline docs/helm-chart/keda.rst:27]: i removed this sentence because if i recall correctly the "we" is astronomer, so in this repo it doesn't really have the same meaning and i don't think the statement belongs.  but it's also sortof beside the point.

File: chart/tests/helm_template_generator.py
Comment (dstandish) [inline chart/tests/helm_template_generator.py:64]: `validate_schema` was added recently to allow tests for CRDs but with this PR we have a way of pulling in the definition.
it does require a network call here fwiw: https://github.com/apache/airflow/pull/13209/files#diff-91361141bde70cbfc90142ebe2377ce918025d3343d9216e00e50c66f083b12cR65

File: chart/templates/workers/worker-kedaautoscaler.yaml
Comment (dstandish) [inline chart/templates/workers/worker-kedaautoscaler.yaml:47]: while updating `connectionFromEnv` just decided to make this a bit more readable by splitting onto separate lines

File: docs/helm-chart/keda.rst
Comment (turbaszek) [reply docs/helm-chart/keda.rst:27]: Exactly it was me and Daniel, so not only Astronomer. Also Airflow is on KEDA users page. But I agree that it creates not value here.

File: chart/README.md
Comment (dstandish) [reply chart/README.md]: Fixed thanks

**Review by ianstanton — APPROVED** [review]

**Comment by dstandish** [comment]
@mik-laj do you want to have another look post-rebase?

Also what is the convention for committer-authors?  Should the author merge it themselves after there is approval or is it expected that you wait for the reviewer to merge?  

Thank you 🙏

**Comment by ashb** [comment]
> Also what is the convention for committer-authors? Should the author merge it themselves after there is approval or is it expected that you wait for the reviewer to merge?

Generally we can merge our selves after a review. Use your judgment if you need a re-review.

**Comment by dstandish** [comment]
> Generally we can merge our selves after a review. Use your judgment if you need a re-review

OK in that case since the rebase was just about docs and nothing materially changed i'll go ahead and merge.

Thanks all

**Comment by dstandish** [comment]
forgot to push fixup adding `autoscaler` to spelling wordlist will merge afterward

File: docs/helm-chart/keda.rst
Comment (mik-laj) [inline docs/helm-chart/keda.rst]: this does not work because it is not valid SQL. It is mix of Jinja and SQL, so we should have 2 lexes here, but it is supported by pygments.

Comment (dstandish) [reply docs/helm-chart/keda.rst]: darn.  yes will fix that too.

Comment (dstandish) [reply docs/helm-chart/keda.rst]: pushed.

**Comment by dstandish** [comment]
@mik-laj @ianstanton  after taking another look at this i thought that adding the namedtuple in this case was maybe a little overkill / confusing because the elements are never accessed it's just a way constructing a key in the dict for the URL lookup, given the two components -- api version and kind.

can you please take a look and let me know what you think of this perhaps simpler (though less explicit) alternative.

and maybe you have a better suggestion.
```python
crd_lookup = {
    'keda.sh/v1alpha1::ScaledObject': 'https://raw.githubusercontent.com/kedacore/keda/v2.0.0/config/crd/bases/keda.sh_scaledobjects.yaml',
}
...
def get_schema_crd(api_version, kind):
    url = crd_lookup.get(f"{api_version}::{kind}")
    if not url:
        return None
    response = requests.get(url)
    yaml_schema = response.content.decode('utf-8')
    schema = yaml.safe_load(StringIO(yaml_schema))
    return schema
```

File: docs/helm-chart/keda.rst
Comment (ashb) [reply docs/helm-chart/keda.rst]: We could set it to `:: jinja` I think, but that probably expects it to be template to output HTML, so would likely be worse than no highlighting.

Comment (ashb) [reply docs/helm-chart/keda.rst]: Yup, tested it (on terminal with `pygmentize -l jinja`) and it looks ugly.

File: chart/tests/helm_template_generator.py
Comment (dstandish) [inline chart/tests/helm_template_generator.py:34]: the concatenation of api version and kind in this way, `keda.sh/v1alpha1::ScaledObject` is a made-up syntax that only serves to provide a way to look up the url for a particular combination of  api version + kind

File: chart/tests/helm_template_generator.py
Comment (dstandish) [inline chart/tests/helm_template_generator.py:37]: the above concatenation of api version and kind in this way, keda.sh/v1alpha1::ScaledObject is a made-up syntax that only serves to provide a simple way to look up the url for a particular combination of api version + kind for CRDs

it is only referenced in one place:
https://github.com/apache/airflow/blob/3805e1ee4d6601956b92130ec7ae6bad9f2243a9/chart/tests/helm_template_generator.py#L57

**Comment by ianstanton** [comment]
> @mik-laj @ianstanton after taking another look at this i thought that adding the namedtuple in this case was maybe a little overkill / confusing because the elements are never accessed it's just a way constructing a key in the dict for the URL lookup, given the two components -- api version and kind.
> 
> can you please take a look and let me know what you think of this perhaps simpler (though less explicit) alternative.
> 
> and maybe you have a better suggestion.
> 
> ```python
> crd_lookup = {
>     'keda.sh/v1alpha1::ScaledObject': 'https://raw.githubusercontent.com/kedacore/keda/v2.0.0/config/crd/bases/keda.sh_scaledobjects.yaml',
> }
> ...
> def get_schema_crd(api_version, kind):
>     url = crd_lookup.get(f"{api_version}::{kind}")
>     if not url:
>         return None
>     response = requests.get(url)
>     yaml_schema = response.content.decode('utf-8')
>     schema = yaml.safe_load(StringIO(yaml_schema))
>     return schema
> ```
> 
> so i'm just joining the two components with `::` to make a string that's pretty straightforward to understand so that we don't need the extra K8sObject namedtuple. Also leaving the key as a regular tuple does work but end up much uglier with line breaks.
> 
> i'm gonna just go ahead and add this in a second commit and request review and go from there

@dstandish this approach seems acceptable to me :+1:

**Comment by dstandish** [comment]
@mik-laj @kaxil  One question about this.... This helm chart isn't "released" ... so I guess there's no need to put anything in `updating.md` for this update at this time.... but let me know if that's not true and you want me to add something somewhere....

**Comment by kaxil** [comment]
> @mik-laj @kaxil One question about this.... This helm chart isn't "released" ... so I guess there's no need to put anything in `updating.md` for this update at this time.... but let me know if that's not true and you want me to add something somewhere....

Yup that is true, currently we don't need to put anything in Updating.md

**Comment by nirroz93** [comment]
Hi, sorry for commenting an old PR. I see here https://github.com/apache/airflow/blob/main/chart/values.yaml#LL501C4-L501C15 there is still a comment 
`# Persistence.enabled must be set to false to use KEDA.`

as far as I understand from this PR, the comment is a leftover (since this PR), right?

**Pull request merged**
